tests/pauli/test_pauli_arithmetic.py

```diff
@@ -0,0 +1,321 @@
+pw4 = PauliWord({})
+
+ps1 = PauliSentence({pw1: 1.23, pw2: 4j, pw3: -0.5})
+ps2 = PauliSentence({pw1: -1.23, pw2: -4j, pw3: 0.5})
+ps3 = PauliSentence({pw3: -0.5, pw4: 1})
+ps4 = PauliSentence({pw4: 1})
+ps5 = PauliSentence({})
+
+
+class TestPauliWord:
+    def test_identity_removed_on_init(self):
+        """Test that identities are removed on init."""
+        pw = PauliWord({0: I, 1: X, 2: Y})
+        assert 0 not in pw.keys()  # identity ops are removed from pw
+
+    def test_missing(self):
+        """Test the result when a missing key is indexed"""
+        pw = PauliWord({0: I, 1: X, 2: Y})
+        assert 3 not in pw.keys()
+        assert pw[3] == I
+
+    def test_set_items(self):
+        """Test that setting items raises an error"""
+        pw = PauliWord({0: I, 1: X, 2: Y})
+        with pytest.raises(NotImplementedError):
+            pw[3] = Z  # trying to add to a pw after instantiation is prohibited
+
+    def test_update_items(self):
+        """Test that updating items raises an error"""
+        pw = PauliWord({0: I, 1: X, 2: Y})
+        with pytest.raises(NotImplementedError):
+            pw.update({3: Z})  # trying to add to a pw after instantiation is prohibited
+
+    def test_hash(self):
+        """Test that a unique hash exists for different PauliWords."""
+        pw_1 = PauliWord({0: I, 1: X, 2: Y})
+        pw_2 = PauliWord({0: I, 1: X, 2: Y})  # same as 1
+        pw_3 = PauliWord({1: X, 2: Y, 0: I})  # same as 1 but reordered
+        pw_4 = PauliWord({1: Z, 2: Z})  # distinct from above
+
+        assert pw_1.__hash__() == pw_2.__hash__()
+        assert pw_1.__hash__() == pw_3.__hash__()
+        assert pw_1.__hash__() != pw_4.__hash__()
+
+    @pytest.mark.parametrize("pw", (pw1, pw2, pw3, pw4))
+    def test_copy(self, pw):
+        """Test that the copy is identical to the original."""
+        copy_pw = copy(pw)
+        assert copy_pw == pw
+
+    tup_pws_wires = ((pw1, {1, 2}), (pw2, {"a", "b", "c"}), (pw3, {0, "b", "c"}), (pw4, set()))
+
+    @pytest.mark.parametrize("pw, wires", tup_pws_wires)
+    def test_wires(self, pw, wires):
+        """Test that the wires are tracked correctly."""
+        assert pw.wires == wires
+
+    tup_pw_str = (
+        (pw1, "[X(1)] @ [Y(2)]"),
+        (pw2, "[X(a)] @ [X(b)] @ [Z(c)]"),
+        (pw3, "[Z(0)] @ [Z(b)] @ [Z(c)]"),
+        (pw4, "[()]"),
+    )
+
+    @pytest.mark.parametrize("pw, str_rep", tup_pw_str)
+    def test_str(self, pw, str_rep):
+        assert str(pw) == str_rep
+
+    tup_pws_mult = (
+        (pw1, pw1, PauliWord({}), 1.0),  # identities are automatically removed !
+        (pw1, pw3, PauliWord({0: Z, 1: X, 2: Y, "b": Z, "c": Z}), 1.0),
+        (pw2, pw3, PauliWord({"a": X, "b": Y, 0: Z}), -1.0j),
+        (pw3, pw4, pw3, 1.0),
+    )
+
+    @pytest.mark.parametrize("pw1, pw2, result_pw, coeff", tup_pws_mult)
+    def test_mul(self, pw1, pw2, result_pw, coeff):
+        copy_pw1 = copy(pw1)
+        copy_pw2 = copy(pw2)
```

Comment:
Why copying the classes? They should NOT be modified during the `__mult__` operation

Reply:
This was old, has been removed since